vowpalwabbit/parse_primitives.cc

```diff
@@ -38,81 +21,52 @@ hash_func_t getHasher(const std::string& s)
     THROW("Unknown hash function: " << s);
 }
 
-bool operator==(const substring& ss, const char* str) { return substring_equal(ss, str); }
-
-bool operator==(const char* str, const substring& ss) { return substring_equal(ss, str); }
-
-bool operator==(const substring& ss1, const substring& ss2) { return substring_equal(ss1, ss2); }
-
-bool operator!=(const substring& ss, const char* str) { return !(ss == str); }
-
-bool operator!=(const char* str, const substring& ss) { return !(ss == str); }
-
-bool operator!=(const substring& ss1, const substring& ss2) { return !(ss1 == ss2); }
-
-std::vector<substring> escaped_tokenize(char delim, substring s, bool allow_empty)
+std::vector<std::string> escaped_tokenize(char delim, VW::string_view s, bool allow_empty)
 {
-  std::vector<substring> tokens;
-  substring current;
-  current.begin = s.begin;
-  bool in_escape = false;
-  char* reading_head = s.begin;
-  char* writing_head = s.begin;
+  std::vector<std::string> tokens;
+  std::string current;
+  size_t end_pos = 0;
+  const char delims[3] = {'\\', delim, '\0'};
+  bool last_space = false;
 
-  while (reading_head < s.end)
+  while (!s.empty() && ((end_pos = s.find_first_of(delims)) != VW::string_view::npos))
   {
-    char current_character = *reading_head++;
-
-    if (in_escape)
+    if(s[end_pos] == '\\')
     {
-      *writing_head++ = current_character;
-      in_escape = false;
-    }
-    else
-    {
-      if (current_character == delim)
-      {
-        current.end = writing_head++;
-        *current.end = '\0';
-        if (current.begin != current.end || allow_empty)
-        {
-          tokens.push_back(current);
-        }
+      current.append(s.begin(), end_pos);
+      s.remove_prefix(end_pos + 1);
 
-        // Regardless of whether the token was saved, we need to reset the current token.
-        current.begin = writing_head;
-        current.end = writing_head;
-      }
-      else if (current_character == '\\')
+      // always insert the next character after an escape if it exists
+      if(!s.empty())
       {
-        in_escape = !in_escape;
+	current.append(s.begin(), 1);
```

Comment:
[Minor][Perf] Do we have a sense of what the performance differences between pre/post change? (It is not used in a hot path, so not necessarily an issue, but I am curious)

Reply:
We lose perf in the escaped tokenize function because we need to copy the strings instead of mutating it in place. The performance for the non-destructive tokenize function should be basically unchanged